src/adapter/src/optimize/subscribe.rs

```diff
@@ -0,0 +1,322 @@
+        transient_id: GlobalId,
+        conn_id: ConnectionId,
+        with_snapshot: bool,
+        up_to: Option<Timestamp>,
+        config: OptimizerConfig,
+    ) -> Self {
+        Self {
+            typecheck_ctx: empty_context(),
+            catalog,
+            compute_instance,
+            transient_id,
+            conn_id,
+            with_snapshot,
+            up_to,
+            config,
+        }
+    }
+}
+
+impl Optimize<SubscribeFrom> for OptimizeSubscribe {
+    type To = GlobalMirPlan<Unresolved>;
+
+    fn optimize(&mut self, plan: SubscribeFrom) -> Result<Self::To, OptimizerError> {
+        let sink_name = format!("subscribe-{}", self.transient_id);
+
+        let (df_desc, df_meta) = match plan {
+            SubscribeFrom::Id(from_id) => {
+                let from = self.catalog.get_entry(&from_id);
+                let from_desc = from
+                    .desc(
+                        &self
+                            .catalog
+                            .state()
+                            .resolve_full_name(from.name(), Some(&self.conn_id)),
+                    )
+                    .expect("subscribes can only be run on items with descs")
+                    .into_owned();
+
+                // Make SinkDesc
+                let sink_id = self.transient_id;
+                let sink_desc = ComputeSinkDesc {
+                    from: from_id,
+                    from_desc,
+                    connection: ComputeSinkConnection::Subscribe(SubscribeSinkConnection::default()),
+                    with_snapshot: self.with_snapshot,
+                    up_to: self.up_to.map(Antichain::from_elem).unwrap_or_default(),
+                    // No `FORCE NOT NULL` for subscribes
+                    non_null_assertions: vec![],
+                };
+
+                let mut df_builder =
+                    DataflowBuilder::new(self.catalog.state(), self.compute_instance.clone());
+
+                let (df_desc, df_meta) =
+                    df_builder.build_sink_dataflow(sink_name, sink_id, sink_desc)?;
+
+                (df_desc, df_meta)
+            }
+            SubscribeFrom::Query { expr, desc } => {
+                // TODO: expr: HirRelationExpr
+                // HIR ⇒ MIR lowering and decorrelation
+                // let expr = expr.lower()?;
+
+                // MIR ⇒ MIR optimization (local)
+                let expr = span!(target: "optimizer", Level::TRACE, "local").in_scope(|| {
+                    let optimizer = TransformOptimizer::logical_optimizer(&self.typecheck_ctx);
+                    let expr = optimizer.optimize(expr)?;
+
+                    // Trace the result of this phase.
+                    trace_plan(&expr);
+
+                    Ok::<_, OptimizerError>(expr)
+                })?;
+
+                // I'm not sure why this is needed - possibly for refining the type?
+                // TODO: figure out if we should be doing this elsewhere.
```

Comment:
Is the question why we need the relation description in the sink description?

Looking at the code it seems like we only need the arity in one place: https://github.com/teskje/materialize/blob/b92239698f48e03a2820f1b95d7e62780a52cce8/src/compute/src/render/sinks.rs#L72